src/main/java/io/confluent/copycat/hdfs/HdfsWriter.java

```diff
@@ -72,95 +95,152 @@ public HdfsWriter(HdfsSinkConnectorConfig connectorConfig) {
     }
   }
 
-
   public Map<TopicPartition, Long> getPreviousOffsets() {
     return offsets;
   }
 
-  public void writeRecord(TopicPartition topicPart, SinkRecord record)
-      throws CopycatException{
+  public void writeRecord(final TopicPartition topicPart, final SinkRecord record)
+      throws CopycatException, InterruptedException {
+    lock.lock();
     try {
+      recover(topicPart);
+
       RecordWriter<Long, SinkRecord> writer = getWriter(topicPart, record);
       writer.write(System.currentTimeMillis(), record);
+
       if (!offsets.containsKey(topicPart)) {
         offsets.put(topicPart, record.getOffset() - 1);
       }
       updateRecordCounter(topicPart);
+
+      if (!tasks.containsKey(topicPart)) {
+        ScheduledFuture task = timedRotateThread.schedule(new Runnable() {
+          @Override
+          public void run() {
+            lock.lock();
+            try {
+              rotate(topicPart);
+            } catch (IOException | InterruptedException e) {
+              throw new CopycatRuntimeException(e);
```

Comment:
Well, CopycatException vs CopycatRuntimeException is removed in the final patch. As discussed offline, we'll introduce some subclasses of CopycatException so connectors can indicate certain types of errors to the framework, e.g. transient vs permanent errors.